resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java

```diff
@@ -327,15 +317,28 @@ boolean clearAndCancel() {
                 return false;
             }
 
-            cancelExpiration(entries);
+            ScheduledFuture<?> expirationFuture = this.expirationFuture;
```

Comment:
`this.expirationFuture` is modified in a synchronized block, but accessed outside a synchronized block here. Can you describe how visibility is guaranteed here and the implications if the change is not visible? 

An alternative approach would be to avoid using `synchronized` and using a token/marker object to indicate cancelled.

```diff
+diff --git a/resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java b/resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java
index 6ce975ff7a..5004929512 100644
--- a/resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java
+++ b/resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java
@@ -17,6 +17,8 @@ package io.netty.resolver.dns;
 
 import io.netty.channel.EventLoop;
 import io.netty.handler.codec.dns.DnsRecord;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
 import io.netty.util.concurrent.ScheduledFuture;
 import io.netty.util.internal.PlatformDependent;
 import io.netty.util.internal.UnstableApi;
@@ -28,7 +30,10 @@ import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicReference;
 
 import static io.netty.util.internal.ObjectUtil.checkNotNull;
@@ -232,8 +237,125 @@ public class DefaultDnsCache implements DnsCache {
 
         private final String hostname;
 
-        // This is guarded by a synchronized block.
-        private ScheduledFuture<?> expirationFuture;
+        // TODO: most methods don't matter, because the only meaningful method is getDelay, but cleanup this code.
+        private final ScheduledFuture<?> CANCELLED = new ScheduledFuture<Object>() {
+            @Override
+            public boolean isSuccess() {
+                return false;
+            }
+
+            @Override
+            public boolean isCancellable() {
+                return false;
+            }
+
+            @Override
+            public Throwable cause() {
+                return null;
+            }
+
+            @Override
+            public Future<Object> addListener(GenericFutureListener<? extends Future<? super Object>> listener) {
+                return null;
+            }
+
+            @Override
+            public Future<Object> addListeners(GenericFutureListener<? extends Future<? super Object>>... listeners) {
+                return null;
+            }
+
+            @Override
+            public Future<Object> removeListener(GenericFutureListener<? extends Future<? super Object>> listener) {
+                return null;
+            }
+
+            @Override
+            public Future<Object> removeListeners(GenericFutureListener<? extends Future<? super Object>>... listeners) {
+                return null;
+            }
+
+            @Override
+            public Future<Object> sync() throws InterruptedException {
+                return null;
+            }
+
+            @Override
+            public Future<Object> syncUninterruptibly() {
+                return null;
+            }
+
+            @Override
+            public Future<Object> await() throws InterruptedException {
+                return null;
+            }
+
+            @Override
+            public Future<Object> awaitUninterruptibly() {
+                return null;
+            }
+
+            @Override
+            public boolean await(long timeout, TimeUnit unit) throws InterruptedException {
+                return false;
+            }
+
+            @Override
+            public boolean await(long timeoutMillis) throws InterruptedException {
+                return false;
+            }
+
+            @Override
+            public boolean awaitUninterruptibly(long timeout, TimeUnit unit) {
+                return false;
+            }
+
+            @Override
+            public boolean awaitUninterruptibly(long timeoutMillis) {
+                return false;
+            }
+
+            @Override
+            public Object getNow() {
+                return null;
+            }
+
+            @Override
+            public boolean cancel(boolean mayInterruptIfRunning) {
+                return false;
+            }
+
+            @Override
+            public long getDelay(TimeUnit unit) {
+                return Long.MIN_VALUE;
+            }
+
+            @Override
+            public int compareTo(Delayed o) {
+                return 0;
+            }
+
+            @Override
+            public boolean isCancelled() {
+                return false;
+            }
+
+            @Override
+            public boolean isDone() {
+                return false;
+            }
+
+            @Override
+            public Object get() throws InterruptedException, ExecutionException {
+                return null;
+            }
+
+            @Override
+            public Object get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+                return null;
+            }
+        };
+        // TODO: make this atomic reference updater
+        private final AtomicReference<ScheduledFuture<?>> expirationFutureRef = new AtomicReference<ScheduledFuture<?>>();
 
         Entries(String hostname) {
             super(Collections.<DefaultDnsCacheEntry>emptyList());
@@ -296,19 +418,29 @@ public class DefaultDnsCache implements DnsCache {
         }
 
         private void scheduleCacheExpirationIfNeeded(int ttl, EventLoop loop) {
-            ScheduledFuture<?> oldScheduled;
-
-            synchronized (this) {
-                oldScheduled = this.expirationFuture;
-                if (oldScheduled == null || oldScheduled.getDelay(TimeUnit.SECONDS) > ttl) {
-                    this.expirationFuture = loop.schedule(this, ttl, TimeUnit.SECONDS);
+            for (;;) {
+                ScheduledFuture<?> oldFuture = expirationFutureRef.get();
+                if (oldFuture == null || oldFuture.getDelay(TimeUnit.SECONDS) > ttl) {
+                    ScheduledFuture<?> newFuture = loop.schedule(this, ttl, TimeUnit.SECONDS);
+                    // It is possible that
+                    // 1. task will fire in between this line, or
+                    // 2. multiple timers may be set if there is concurrency
+                    // (1) Shouldn't be a problem because we will fail the CAS and then the next loop will see CANCELLED
+                    //     so the ttl will not be less, and we will bail out of the loop.
+                    // (2) This is a trade-off to avoid concurrency resulting in contention on a synchronized block.
+                    //     IIUC this would be an overall reduction in existing scheduling so this is likely acceptable.
+                    if (expirationFutureRef.compareAndSet(oldFuture, newFuture)) {
+                        if (oldFuture != null) {
+                            oldFuture.cancel(true);
+                        }
+                        break;
+                    } else {
+                        newFuture.cancel(true);
+                    }
                 } else {
-                    oldScheduled = null;
+                    break;
                 }
             }
-            if (oldScheduled != null) {
-                oldScheduled.cancel(false);
-            }
         }
 
         boolean clearAndCancel() {
@@ -317,7 +449,7 @@ public class DefaultDnsCache implements DnsCache {
                 return false;
             }
 
-            ScheduledFuture<?> expirationFuture = this.expirationFuture;
+            ScheduledFuture<?> expirationFuture = expirationFutureRef.getAndSet(CANCELLED);
             if (expirationFuture != null) {
                 expirationFuture.cancel(false);
             }
```

Note existing unit tests pass but more review would be useful :)

Reply:
@Scottmitch marked as volatile for now to fix the visibility concerns (good catch). Let me think about your idea a bit. I guess at worse we could do this as a followup ? IMHO the existing approach in this PR is already a nice improvement. Also note that I would not expect a lot of condition on the synchronized block here as the synchronisation only happens when modification for the same hostname happens concurrently (which most likely will not happened).